src/pymor/analyticalproblems/functions.py

```diff
@@ -61,22 +61,72 @@ def __add__(self, other):
+            else:
+                functions, coefficients = (self, other), (1., sign)
+        elif other.name == 'LincombFunction' and isinstance(other, LincombFunction):
+            functions = self.functions + other.functions
+            coefficients = self.coefficients + (other.coefficients if sign == 1.
+                                                else tuple(-c for c in other.coefficients))
+        else:
+            functions, coefficients = self.functions + (other,), self.coefficients + (sign,)
 
-    def __sub__(self, other):
-        if isinstance(other, Function):
-            return LincombFunction([self, other], [1., -1.])
+        return LincombFunction(functions, coefficients)
+
+    def _radd_sub(self, other, sign):
+        if isinstance(other, Number) and other == 0:
+            return self
+        elif not isinstance(other, Function):
+            other = np.array(other)
+            assert other.shape == self.shape_range
+            if np.all(other == 0.):
+                return self
+            other = ConstantFunction(other, dim_domain=self.dim_domain)
+
+        # note that 'other' can never be a LincombFunction
+        if self.name != 'LincombFunction' or (self.name == 'LincombFunction'
+                                              and not isinstance(self, LincombFunction)):
+            functions, coefficients = (other, self), (1., sign)
         else:
-            return self + (- np.array(other))
+            functions = (other,) + self.functions
+            coefficients = (1.,) + (self.coefficients if sign == 1. else tuple(-c for c in self.coefficients))
+
+        return LincombFunction(functions, coefficients)
+
+    def __add__(self, other):
+        return self._add_sub(other, 1.)
+
+    def __sub__(self, other):
+        return self._add_sub(other, -1.)
+
+    def __radd__(self, other):
+        return self._radd_sub(other, 1.)
+
+    def __rsub__(self, other):
+        return self._radd_sub(other, -1.)
 
     def __mul__(self, other):
+        if not isinstance(other, (Number, ParameterFunctional, Function)):
+            return NotImplemented
         if isinstance(other, (Number, ParameterFunctional)):
             return LincombFunction([self], [other])
-        if isinstance(other, Function):
-            return ProductFunction([self, other])
-        return NotImplemented
+        if self.name != 'ProductFunction':
+            if isinstance(other, ProductFunction) and other.name == 'ProductFunction':
+                return other.with_(functions=other.functions + [self])
+            else:
+                return ProductFunction([self, other])
+        elif isinstance(other, ProductFunction) and other.name == 'ProductFunction':
+            assert isinstance(self, ProductFunction)
+            functions = self.functions + other.functions
+            return ProductFunction(functions)
+        else:
+            assert isinstance(self, ProductFunction)
```

Comment:
I guess that here we have in principle the same issue as with `LincombFunction` and probably should not be so sure that we actually have a `ProductFunction`.

Reply:
Yes, I missed this part : /